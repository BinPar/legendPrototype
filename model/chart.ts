/**
 * Information about label
 */
export interface LabelInfo {
  /** Unique ID of the label */
  key: string;
  /** Text to display on the label */
  name: string;
  /** X coordinate of the interest point */
  cx: number;
  /** Y coordinate of the interest point */
  cy: number;
  /** X coordinate of the label start */
  tx: number;
  /** Y coordinate of the label start */
  ty: number;
  /** The with of the label in pixels */
  labelWidth: number;  
  /**
   * Property that will force the ElbowMode of the line drawing that will use an 
   * L shape to connect the label with the point instad of a direct line.
   * This shape will fix some problems in the drawing
   */
  useElbowMode: boolean;
  /** Label position (right or left) */
  labelPosition: 'right' | 'left';
  /** True when the line is crossed by another line */
  isCrossed: boolean;
}

/**
 * Settings of the current chart
 */
export interface ChartProps {
  /**
   * Seed of the chart: the same seed will generate exactly the same random result
   * similar to how it works in MineCraft in order to allow to share the URL of any
   * unusual combination
   */
  seed?: string;
}
