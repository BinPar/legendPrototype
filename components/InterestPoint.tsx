import React from 'react';
import { LabelInfo } from '../model/chart';

/**
 * Returns the SVG line of the point of interest and other SVG visual details
 * @param {LabelInfo} point Information about the point to draw in the SVG
 * @returns SVG group with the visual elements to draw the interest point
 */
const InterestPoint = (point: LabelInfo): JSX.Element => (
  <g key={`${point.key}`} filter="url(#sofGlow)">
    <line
      x1={point.cx}
      y1={point.cy}
      x2={point.tx}
      y2={point.ty}
      stroke={point.isCrossed ? 'red' : 'black'}
      strokeWidth="3"
    />
    <circle
      cx={point.cx}
      cy={point.cy}
      r={5}
      fill="#000"
      filter="url(#sofGlow)"
    />
  </g>
);

export default InterestPoint;
