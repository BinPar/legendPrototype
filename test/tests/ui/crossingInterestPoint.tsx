import React from 'react';
import { mount } from 'enzyme';
import InterestPoint from '../../../components/InterestPoint';
import { LabelInfo } from '../../../model/chart';

/**
 * Checks the draw of interest points where they
 * have crossing lines
 */
test('Crossing interest point', () => {

  /**
   * Mocked draw info
   */
  const props: LabelInfo = {
    key: 'test',
    name: 'test',
    labelWidth: 20,
    labelPosition: 'left',
    isCrossed: true,
    cx: 40,
    cy: 70,
    tx: 10,
    ty: 20,
  };

  const lines = [props];

  // Renders the InterestPoint
  const result = mount(
    <svg width={1024} height={720}>
      {lines.map(InterestPoint)}
    </svg>
  );

  // Verify that the Virtual DOM is not changed
  expect(result).toMatchSnapshot();
});
