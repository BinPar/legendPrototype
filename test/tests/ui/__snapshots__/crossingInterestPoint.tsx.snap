// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crossing interest point: crossedInterestPoint 1`] = `
<svg
  height={720}
  width={1024}
>
  <g
    filter="url(#sofGlow)"
    key="test"
  >
    <line
      stroke="red"
      strokeWidth="3"
      x1={40}
      x2={10}
      y1={70}
      y2={20}
    />
    <circle
      cx={40}
      cy={70}
      fill="#000"
      r={5}
    />
  </g>
</svg>
`;

exports[`Crossing interest point: crossedInterestPointWithElbow 1`] = `
<svg
  height={720}
  width={1024}
>
  <g
    filter="url(#sofGlow)"
    key="test"
  >
    <line
      stroke="red"
      strokeWidth="3"
      x1={10}
      x2={40}
      y1={20}
      y2={20}
    />
    <line
      stroke="red"
      strokeWidth="3"
      x1={40}
      x2={40}
      y1={70}
      y2={20}
    />
    <circle
      cx={40}
      cy={70}
      fill="#000"
      r={5}
    />
  </g>
</svg>
`;
